File: morning_exercises/day13.py
from __future__ import annotations

class BSTNode:
    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None
    
    @staticmethod
    def Validator(root: BSTNode | None, min: float=float('-inf'), max: float=float('inf'))->bool:
        if (root is None): return True
        if (root.val>=max or root.val<=min): return False
        return (BSTNode.Validator(root.left, min, root.val) and BSTNode.Validator(root.right, root.val, max))
    
root = BSTNode(6)
root.left = BSTNode(4)
root.right = BSTNode(9)
root.left.left = BSTNode(2)
root.left.right = BSTNode(5)
root.right.left = BSTNode(7)
root.right.right = BSTNode(10)

print(BSTNode.Validator(root))

File: morning_exercises/day13.cpp
#include <climits>
#include <ios>
#include <memory>
#include <iostream>

class BSTNode
{
    public:
        int val;
        std::shared_ptr<BSTNode> left;
        std::shared_ptr<BSTNode> right;
        
    BSTNode(int val = 0): val(val), left(nullptr), right(nullptr) {}

    static bool isValidBST(std::shared_ptr<BSTNode> root, long min, long max) {
        if (root==nullptr) return true;
        if (root->val>=max || root->val<=min) return false;
        return (isValidBST(root->left, min, root->val) && isValidBST(root->right, root->val, max));
    }

    static bool validation(std::shared_ptr<BSTNode> root) {
        return isValidBST(root, LONG_MIN, LONG_MAX);
    }
};

int main() {
    auto root = std::make_shared<BSTNode>(5);
    root->left = std::make_shared<BSTNode>(3);
    root->right = std::make_shared<BSTNode>(7);
    root->left->left = std::make_shared<BSTNode>(2);
    root->left->right = std::make_shared<BSTNode>(4);
    std::cout <<std::boolalpha;
    std::cout << BSTNode::validation(root) << "\n";
}